### DISCUSSION TITLE
OME WebRTC stream drifting away from realtime

### DISCUSSION THREAD

**Discussion by devedse** [OvenMediaLabs/OvenMediaEngine#1275] [Help]
**Describe the bug**
I'm currently using OpenStreamingPlatform to stream games together with my friends. However over the span of about 30-minutes+ the stream starts being less and less realtime. Initially there's about 0.1-0.2 seconds delay but later this becomes 4 seconds or more.

If you refresh the page however, the delay is instantly gone.

**To Reproduce**
* All my settings are written down in this Discussion: https://github.com/AirenSoft/OvenMediaEngine/discussions/1238

**Expected behavior**
I would expect the stream to stay live as long as I'm watching

**Logs**
See: https://github.com/AirenSoft/OvenMediaEngine/discussions/1238

**Server (please complete the following information):**
 - OS: Raspberry PI 4 ARM 64 + docker
 - OvenMediaEngine Version: [0.15.12](https://hub.docker.com/layers/airensoft/ovenmediaengine/0.15.12/images/sha256-5858ba5a752dd0da2ad5d9c0569f0feebea3dad7425eca0f0da354b06dbb1824?context=explore)
 - Branch: N/A

**Player (please complete the following information):**
 - Device: Windows PC
 - OS: Windows 11
 - Browser: Chrome
 - Version: Latest

**Additional context**
N/A

**Comment by getroot** [comment]
This is most likely not an issue with OvenMediaEngine. Or do you have reason to believe this is a bug or issue in OME? If so, please share that information.

It could be if the network speed between OME and Player is not fast enough and you are playing with WebRTC/TCP. (because TCP retransmits)

Try lowering the bitrates of the output and playing it. Or try playing with WebRTC/UDP.

**Reply by devedse** [reply]
I'm currently streaming locally and this still happens (so over my own LAN). It does indeed happen quicker to people watching my streams externally. My bitrate is set to 1000kbps (really low I know) but it still happens.

However I would love if there would be some feature that would "catch up".

For me when I'm interacting with my viewers it's very frustrating that there's a 10 second delay after a while. And having to ask my viewers to keep refreshing the page very 15-30 minutes is a bit of an inconvenience.

**Reply by getroot** [reply]
The problem is not reproducible in my environment. (I have several production environments and have never had such a problem except for the player's network speed.)

I have no experience with Raspberry Pi. Is that server CPU usage high enough? Did you do any kernel tuning? https://airensoft.gitbook.io/ovenmediaengine/troubleshooting#5-2.-tuning-your-linux-kernel

**Reply by devedse** [reply]
I could take a look into this, however I found expect that even during periods of higher latency it should be able to catch up later.

I did try 30.000 (30mbit) streams as well previously in my old home where my internet was better, and these had the same problem in the long run. But not exponentially more or anything. So the Raspberry PI can at least handle the load (even to multiple users). It's just that I experience the delay the longer I watch.

Have you tried running a stream with a timer on the screen for long periods of time to see if there's any drift at all?

**Reply by getroot** [reply]
Of course, OME has been playing hundreds of hours in my clients' production environments. We also played a BIPBOB video over and over during a show at NAB a while ago.

> If you refresh the page however, the delay is instantly gone.

means, 

- The server kernel's socket buffer is getting bigger: Server performance or network speed is insufficient or the buffer is insufficient
- WebRTC player (browser) jitter buffer is getting bigger: A/V sync is getting out of sync or network jitter is getting worse

What is sending a stream to OME? Do you use OBS?? Or is it another system from OpenStreamingPlatform?

**Reply by getroot** [reply]
Are your encoder settings still as below? 
https://github.com/AirenSoft/OvenMediaEngine/discussions/1238#discussioncomment-5945903

Please note that WebRTC does not support AVC High Profile. Check if the problem is reproduced with the WebRTC recommended profile.
https://airensoft.gitbook.io/ovenmediaengine/quick-start#publishing

**Reply by getroot** [reply]
I ran the test for over 30 minutes to check for possible regression bugs, but the latency starts with 8 frames and doesn't increase anymore. I will run this longer.

![image](https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/2a5d1ace-7304-4ae6-ae2b-2aa1b22c5d2a)
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/d4d2c308-74f9-48a6-8ea8-147b70284300)

**Reply by devedse** [reply]
I'll take some time to further investigate this later :), thanks so much for the instant elaborate responses.

**Reply by devedse** [reply]
I've just started a stream which I'm watching on my local network. I am indeed using OpenStreamingPlatform as you can see but I would expect the problem to lie with OME, the reasoning being that if I refresh the page the delay goes away again.

Start of the stream the delay is about 0.29 seconds:
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/2350015/37a69606-b766-45b7-aeb8-4da8418bd996)

After about 5 minutes the delay is 0.45 seconds:
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/2350015/c5c25c31-8dab-461b-838b-4b9876e2d763)

And after 10 minutes the delay is 0.48:
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/2350015/508e4ba4-b39c-45f2-8248-6a87fbd1f279)

![image](https://github.com/AirenSoft/OvenMediaEngine/assets/2350015/cc185d5d-c715-4749-baa9-de2860d7221b)

I'll leave it running a bit longer to see what happens. I do think you're right that the problem is worse when people are watching the stream over the internet.

My question is though, let's say the connection drops for about 3 seconds. Does that mean the player will be "behind" 3 seconds for ever and ever until you refresh the page?

**Reply by devedse** [reply]
I did indeed see that the Profile was set to `high` rather then `baseline`. I'll try to update that to see if it makes a difference. Here's my settings now:

![image](https://github.com/AirenSoft/OvenMediaEngine/assets/2350015/066f6e00-0116-4e46-b625-c54beb4d7748)

**Reply by devedse** [reply]
I started a new stream with baseline set to main:

0.35 delay initially: (I didn't reset the timer)
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/2350015/8d3cdea4-a5cd-4c19-8859-1e87db677f2a)

0.45 about 6 minutes later:
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/2350015/b13e2f7b-91ee-42ca-9abc-5cf22e4f22ba)

0.54 about 30 minutes later:
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/2350015/2a6119c8-a39c-4f6b-a293-3838eeb66c24)

![image](https://github.com/AirenSoft/OvenMediaEngine/assets/2350015/c76b76e2-3f80-4921-bd6c-10e7a4889cb6)

**Reply by getroot** [reply]
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/004c0059-1dc9-4199-85e3-6b95d3f564b2)

Your browser's jitter buffer has increased, and 0.54 seconds seems to be an acceptable latency that can happen depending on network jitter. Looking at the graph, it seems that your jitterbuffer size converges around that. For reference, Chrome's jitter buffer is a value that Chrome monitors various conditions and dynamically adjusts internally, so it cannot be controlled externally.

After more time (how much?) in this state, does the latency continue to increase above 4 seconds?

**Reply by devedse** [reply]
I tried it yesterday with a friend of mine for about an hour but we couldn't get the latency significantly higher. I'll do some more research into this and will come back to you with new results over the next few days.

**Reply by getroot** [reply]
Yes, I hope the information below will help you.

- If you play with WebRTC/TCP on a device with a slow network, you won't experience any stuttering, but the latency will continue to increase.

- If you play with WebRTC/UDP on a device with a slow network, the latency won't increase but you will experience stuttering.

**Reply by devedse** [reply]
That makes sense, I'll keep you posted :).

Wouldn't it be possible though to increase the playspeed of WebRTC/TCP when the buffer get's full to be able to catch up? Or is that logic also hidden in Chromium?

**Comment by getroot** [comment]
And you wrote OME RTSP in the title of this issue, is OME WebRTC correct? If so, please edit the title.

**Reply by devedse** [reply]
You're fully correct, I've modified the title
